### ISSUE TITLE
meltedblocks - Missing Slippage Protection in Request Type of Deposit and Redeem

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-08-midas-minter-redeemer-judging#17]
meltedblocks

Medium

# Missing Slippage Protection in Request Type of Deposit and Redeem

### Summary

The missing slippage check in `depositRequest` and `redeemRequest` can lead to losses for users, as provided tokens can be exchanged to mToken with unfavorable ratio.

It should be noted that the Midas team consider following issue as invalid:
_Fees and exchange rates: the fees/exchange rates that are evolving between the standard request, and the processing are not a bug._

This bug is different than this exclusion. Described exchange rate difference happens before the `request` is created, not between request and approval.

### Root Cause

The [depositRequest](https://github.com/sherlock-audit/2024-08-midas-minter-redeemer/blob/main/midas-contracts/contracts/DepositVault.sol#L148) (`DepositVault`) and [redeemRequest](https://github.com/sherlock-audit/2024-08-midas-minter-redeemer/blob/main/midas-contracts/contracts/RedemptionVault.sol#L194) (`RedemptionVault`) was introduced as two-step process to allow token deposits and redeem after daily limit was reached. After user create deposit or redeem request, the `vaultAdmin` can approve or cancel request. 

There is some protection introduced in function `_requireVariationTolerance` and parameter `variationTolerance`. Protocol guarantees that exchange ratio difference on request creation and request approval cannot exceed provided `variationTolerance` limit. 

However, it should be noted that this protection acts only after request is created by user - so after `redeemRequest` and `depositRequest` functions are processed. There is still slippage problem within the time when user do the transaction for `redeemRequest` or `depositRequest` and the moment when these transactions are processed and included in block. Within this time exchange ratio can be dropped to the levels that are not acceptable for users. 

### Internal pre-conditions

User needs to call `depositRequest` (`DepositVault`) and `redeemRequest` (`RedemptionVault`) - no other preconditions are required.

### External pre-conditions

1. Exchange ratio for mToken and USD should drop after user intimates transaction and transaction is included in block.

### Attack Path

1. User initiates redeem or deposit by calling `depositRequest` (`DepositVault`) or `redeemRequest` (`RedemptionVault`). User is expecting some exchange ratio.
2. Before transaction is processed - the expected mToken/USD ratio is modified because of external factors (ratio is provided by oracle). 
3. In the `depositRequest` (`DepositVault`) or `redeemRequest` (`RedemptionVault`) - the request is saved with new modified ratio which is infavor to user.
4. `vaultAdmin` processes the transaction, transaction pass internal checks for  `variationTolerance`.
5. user receives less tokens than expected, because initial ratio in the request was different that user expected.

### Impact

User suffer loss as his tokens are exchanged in unfavourable ratio.

### Mitigation

Users should be able to set `minExchangeRatio` along with the invocation of `depositRequest` and `redeemRequest` to be protected against exchange ratio change.

Duplicate of #104
